arch/riscv/core/reset.S

```diff
@@ -38,17 +41,66 @@ SECTION_FUNC(reset, __reset)
  * the C domain
  */
 SECTION_FUNC(TEXT, __initialize)
-	csrr a0, mhartid
-	beqz a0, boot_first_core
-
-	li t0, CONFIG_MP_MAX_NUM_CPUS
-	blt a0, t0, boot_secondary_core
 
-loop_unconfigured_cores:
+    csrr a0, mhartid /* Who are we? */
+	/* First stage of wake up - we assume for MPFS and other systems with
+	 * monitor hart that array has CONFIG_RV_MP_TOTAL_NUM_CPUS */
+	slli a1, a0, RV_REGSHIFT
+	la a2, hart_wake_flags
+	add a2, a2, a1
+	li a1, RV_WAKE_INIT
+	sr a1, 0(a2)
+#if CONFIG_HAS_MONITOR_HART
+/*
+ * For SMP operation the monitor hart cannot be part of the SMP group
+ * of harts as it is generally not the same as the application harts...
+ * Just loop for ever for now if we get here.
+ */
+	li a1, CONFIG_MONITOR_HART_ID
+	bne a0, a1, not_monitor_hart
+	/*
+	 * Disable interrupts etc to try and keep things quiet.
+	 * But first assign the first non zephyr isr stack to us for safety,
+	 * If monitor hart is not hart id 0, this will need changing...
+	 */
+	la sp, _non_z_stacks
+	li t0, __z_interrupt_stack_SIZEOF
+	add sp, sp, t0
+no_monitor_loop:
+	csrw mstatus, zero
+    csrw mie, zero
+    csrw mip, zero
 	wfi
-	j loop_unconfigured_cores
+	j no_monitor_loop
+
+not_monitor_hart:
+#endif	/* CONFIG_HAS_MONITOR_HART */
+    /*
+     * This will boot primary core which is usually thelowest numbered, just halt
+	 * other cores.
+     *
+     * Note. For systems with monitor hart we assume no supervisor supported on monitor
+	 * hart so if this is selected as the hart to use we don't touch mxdeleg registers.
+	 * Also, if not in master mode this is a bad thing to do...
+     */
+#if CONFIG_NO_SUPERVISOR_DELEGATION
+#if CONFIG_HAS_MONITOR_HART && !CONFIG_MONITOR_HART_SUPERVISOR
+	beq a0, a1, not_super_capable /* regs set up earlier... */
+#endif /* CONFIG_HAS_MONITOR_HART && !CONFIG_MONITOR_HART_SUPERVISOR */
+    csrw mideleg,0
+    csrw medeleg,0
+	/*
+	 * Enable all counters for user mode otherwise some user mode samples
+	 * will fail on real hardware...
+	 */
+	li t0,0xFFFFFFFF
```

Comment:
Can you use the `BIT_MASK` macro here? Also, your changes could use better indentation

Reply:
I used an explicit value here as although I've decided to enable all the counters, it might be tuned to only include some of them depending on what's really required.

I'm not sure what you mean about the indentation?

Reply:
For example, line 93 has a different indentation style than line 99 in the snippet above.

Reply:
Thanks for the clarification, I'll look into it with Conor.